metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -129,155 +149,402 @@ func getResource(resourceName string) kubernetes.Resource {
 	}
 }
 
-// NewResourceMetadataEnricher returns an Enricher configured for kubernetes resource events
+// getExtraWatchers returns a list of the extra resources to watch based on some resource.
+// The full list can be seen in https://github.com/elastic/beats/issues/37243, at Expected Watchers section.
+func getExtraWatchers(resourceName string, config *kubernetesConfig) []string {
+	switch resourceName {
+	case PodResource:
+		extra := []string{NamespaceResource, NodeResource}
+		// We need to create watchers for ReplicaSets and Jobs that it might belong to,
+		// in order to be able to retrieve 2nd layer Owner metadata like in case of:
+		// Deployment -> Replicaset -> Pod
+		// CronJob -> job -> Pod
+		if config.AddResourceMetadata != nil && config.AddResourceMetadata.Deployment {
+			extra = append(extra, ReplicaSetResource)
+		}
+		if config.AddResourceMetadata != nil && config.AddResourceMetadata.CronJob {
+			extra = append(extra, JobResource)
+		}
+		return extra
+	case ServiceResource:
+		return []string{NamespaceResource}
+	case DeploymentResource:
+		return []string{NamespaceResource}
+	case ReplicaSetResource:
+		return []string{NamespaceResource}
+	case StatefulSetResource:
+		return []string{NamespaceResource}
+	case DaemonSetResource:
+		return []string{NamespaceResource}
+	case JobResource:
+		return []string{NamespaceResource}
+	case CronJobResource:
+		return []string{NamespaceResource}
+	case PersistentVolumeResource:
+		return []string{}
+	case PersistentVolumeClaimResource:
+		return []string{NamespaceResource}
+	case StorageClassResource:
+		return []string{}
+	case NodeResource:
+		return []string{}
+	case NamespaceResource:
+		return []string{}
+	default:
+		return []string{}
+	}
+}
+
+// getWatchOptions builds the kubernetes.WatchOptions{} needed for the watcher based on the config and nodeScope
+func getWatchOptions(config *kubernetesConfig, nodeScope bool, client k8sclient.Interface, log *logp.Logger) (*kubernetes.WatchOptions, error) {
+	var err error
+	options := kubernetes.WatchOptions{
+		SyncTimeout: config.SyncPeriod,
+		Namespace:   config.Namespace,
+	}
+
+	// Watch objects in the node only
+	if nodeScope {
+		nd := &kubernetes.DiscoverKubernetesNodeParams{
+			ConfigHost:  config.Node,
+			Client:      client,
+			IsInCluster: kubernetes.IsInCluster(config.KubeConfig),
+			HostUtils:   &kubernetes.DefaultDiscoveryUtils{},
+		}
+		options.Node, err = kubernetes.DiscoverKubernetesNode(log, nd)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't discover kubernetes node: %w", err)
+		}
+	}
+	return &options, err
+}
+
+// startWatcher starts a watcher for a specific resource
+func startWatcher(
```

Comment:
This function does not start the watchers, rather it creates them. It should be renamed to createWatcher

Reply:
I will refactor